File: julia/3d/SpectralUtils.jl
"""
Dimension-agnostic spectral building blocks shared by the 3D Spectral fPINN:
Chebyshev-Gauss-Lobatto (CGL) grid + D2 differentiation matrix, and the
L1 Toeplitz quadrature matrix for the Caputo fractional derivative.

Direct Julia port of spectral_utils.py. Uses Float64 throughout (the Python
version used Float32); this is a deliberate, noted difference for the
cross-validation exercise, not an attempt at bit-exact reproduction.
"""
module SpectralUtils

using LinearAlgebra
using SpecialFunctions: gamma, loggamma

export build_chebyshev_grid_and_d2, build_l1_toeplitz_matrix, build_differentiable_l1_matrix

"""
Constructs shifted Chebyshev-Gauss-Lobatto (CGL) points on [0, L]
and the second-order spectral differentiation matrix D2.
Direct port of spectral_utils.build_chebyshev_grid_and_d2.
"""
function build_chebyshev_grid_and_d2(N::Int, L::Float64=1.0)
    j = collect(0:N)
    xi = cos.(π .* j ./ N)
    x_cgl = 0.5 * L .* (1.0 .- xi)

    c = ones(Float64, N + 1)
    c[1] = 2.0
    c[end] = 2.0
    c .*= (-1.0) .^ j

    X = xi * ones(1, N + 1)
    dX = X - X' + Matrix{Float64}(I, N + 1, N + 1)

    D = (c * (1.0 ./ c)') ./ dX
    D = D - Diagonal(vec(sum(D, dims=2)))

    Dx = (-2.0 / L) .* D
    D2 = Dx * Dx

    return x_cgl, D2
end

"""
Lower-triangular Toeplitz L1-scheme weight matrix for the Caputo derivative
of order gamma in (0,1), with a FIXED (non-differentiable) gamma. Used for
the forward problem. Direct port of spectral_utils.build_l1_toeplitz_matrix.
"""
function build_l1_toeplitz_matrix(Nt::Int, gamma_val::Float64, dt::Float64)
    w = [(Float64(i))^(1.0 - gamma_val) - (Float64(i - 1))^(1.0 - gamma_val) for i in 1:Nt]
    W = zeros(Float64, Nt, Nt)
    for n in 1:Nt
        for m in 1:n
            W[n, m] = w[n - m + 1]
        end
    end
    prefactor = dt^(-gamma_val) / gamma(2.0 - gamma_val)
    return prefactor .* W
end

"""
Same L1 Toeplitz weight matrix, but built via a non-mutating comprehension
from a scalar `gamma_val` that may be a Dual (ForwardDiff) or tracked
(Zygote) number, so gradients flow through the fractional order itself.
Direct port of spectral_utils.build_differentiable_l1_matrix.
"""
function build_differentiable_l1_matrix(Nt::Int, gamma_val, dt::Float64)
    w = [(Float64(i))^(1.0 - gamma_val) - (Float64(i - 1))^(1.0 - gamma_val) for i in 1:Nt]
    W = [m <= n ? w[n - m + 1] : zero(gamma_val) for n in 1:Nt, m in 1:Nt]
    prefactor = dt^(-gamma_val) / exp(loggamma(2.0 - gamma_val))
    return prefactor .* W
end

end # module

File: 3d/spectral_utils.py
"""
Dimension-agnostic spectral building blocks shared by the 3D Spectral fPINN:
Chebyshev-Gauss-Lobatto (CGL) grid + D2 differentiation matrix, and the
L1 Toeplitz quadrature matrix for the Caputo fractional derivative.
"""

import numpy as np
import torch
from scipy.special import gamma as scipy_gamma


def build_chebyshev_grid_and_d2(N, L=1.0):
    """
    Constructs shifted Chebyshev-Gauss-Lobatto (CGL) points on [0, L]
    and the second-order spectral differentiation matrix D2.
    """
    j = np.arange(N + 1)
    xi = np.cos(np.pi * j / N)
    x_cgl = 0.5 * L * (1.0 - xi)

    c = np.ones(N + 1)
    c[0] = 2.0
    c[N] = 2.0
    c *= (-1.0) ** j

    X = np.outer(xi, np.ones(N + 1))
    dX = X - X.T + np.eye(N + 1)

    D = np.outer(c, 1.0 / c) / dX
    D -= np.diag(np.sum(D, axis=1))

    D_x = (-2.0 / L) * D
    D2 = D_x @ D_x

    return torch.tensor(x_cgl, dtype=torch.float32), torch.tensor(D2, dtype=torch.float32)


def build_l1_toeplitz_matrix(N_t, gamma, dt):
    """
    Constructs the lower-triangular Toeplitz weight matrix for the L1 scheme
    of the Caputo fractional derivative of order gamma in (0, 1).

    (D_t^gamma U)_n = prefactor * sum_{k=0}^{n-1} w_k * (U_{n-k} - U_{n-k-1})
    where w_k = (k+1)^(1-gamma) - k^(1-gamma).
    """
    k = np.arange(N_t)
    w = (k + 1.0) ** (1.0 - gamma) - (k) ** (1.0 - gamma)

    W = np.zeros((N_t, N_t), dtype=np.float32)
    for n in range(N_t):
        W[n, :n + 1] = w[:n + 1][::-1]

    prefactor = (dt ** (-gamma)) / scipy_gamma(2.0 - gamma)
    W = prefactor * W
    return torch.tensor(W, dtype=torch.float32)


def build_differentiable_l1_matrix(N_t, gamma_tensor, dt):
    """
    Same L1 Toeplitz weight matrix as build_l1_toeplitz_matrix, but built from
    native PyTorch ops on a trainable gamma_tensor so Autograd can flow through
    the fractional order itself (used for inverse parameter discovery).
    """
    k = torch.arange(N_t, dtype=torch.float32, device=gamma_tensor.device)
    w = (k + 1.0) ** (1.0 - gamma_tensor) - (k) ** (1.0 - gamma_tensor)

    rows = []
    for n in range(N_t):
        row = torch.zeros(N_t, device=gamma_tensor.device)
        row[:n + 1] = torch.flip(w[:n + 1], dims=[0])
        rows.append(row)
    W = torch.stack(rows)

    prefactor = (dt ** (-gamma_tensor)) / torch.exp(torch.special.gammaln(2.0 - gamma_tensor))
    return prefactor * W
